util/conditions/patch.go

```diff
@@ -0,0 +1,140 @@
+	RemoveConditionPatch PatchOperation = "Remove"
+)
+
+// ConditionPatch define a patch that apply to a condition.
+type ConditionPatch struct {
+	Target *clusterv1.Condition
+	Base   *clusterv1.Condition
+	Op     PatchOperation
+}
+
+// GetPatch gets returns the list of ConditionPatch required to align source to target.
+func GetPatch(base clusterv1.Conditions, target clusterv1.Conditions) []ConditionPatch {
+	var patch []ConditionPatch
+
+	// Identify AddCondition and ModifyCondition changes-
+	for i := range target {
+		targetCondition := target[i]
+
+		currentCondition := get(base, targetCondition.Type)
+		if currentCondition == nil {
+			patch = append(patch, ConditionPatch{Op: AddConditionPatch, Target: &targetCondition})
+			continue
+		}
+
+		if !reflect.DeepEqual(&targetCondition, currentCondition) {
+			patch = append(patch, ConditionPatch{Op: ChangeConditionPatch, Target: &targetCondition, Base: currentCondition})
+		}
+	}
+
+	// Identify RemoveCondition changes-
+	for i := range base {
+		currentCondition := base[i]
+
+		targetCondition := get(target, currentCondition.Type)
+		if targetCondition == nil {
+			patch = append(patch, ConditionPatch{Op: RemoveConditionPatch, Base: &currentCondition})
+		}
+	}
+	return patch
+}
+
+// ApplyPatch executes a three way merge of a list of ConditionPatch.
+// When merge conflict are detected (source deviated from the original base recorded when creating the patch
+// in an incompatible way), an error is returned.
+func ApplyPatch(source clusterv1.Conditions, patch []ConditionPatch) (clusterv1.Conditions, error) {
+	merge := source.DeepCopy()
+
+	for i := range patch {
+		conditionPatch := patch[i]
+
+		switch conditionPatch.Op {
+		case AddConditionPatch:
+			// if the condition is already on source, check if source and target agree on the change;
+			// otherwise this is a conflict
+			if sourceCondition := get(source, conditionPatch.Target.Type); sourceCondition != nil {
+				// if source and target agree on the change, then it is a conflict
+				if !hasSameState(sourceCondition, conditionPatch.Target) {
+					return nil, errors.Errorf("error patching conditions: The condition %q on was modified by a different process and this caused a merge conflict", conditionPatch.Target.Type)
+				}
+				// otherwise source is already ok;
+				// NOTE: we are preserving LastTransitionTime from source in order to not altering the existing value.
+				merge = set(merge, sourceCondition)
+				continue
+			}
+			// if the condition does not exists on source, add the new target condition
+			merge = set(merge, conditionPatch.Target)
+
+		case ChangeConditionPatch:
+			sourceCondition := get(source, conditionPatch.Target.Type)
+
+			// if the conditions does not exists anymore on source, this is conflict
+			if sourceCondition == nil {
+				return nil, errors.Errorf("error patching conditions: The condition %q on was modified by a different process and this caused a merge conflict", conditionPatch.Target.Type)
```

Comment:
nit: this error should likely say be updated to indicate that we were attempting to modify a condition that no longer exists.

Reply:
done